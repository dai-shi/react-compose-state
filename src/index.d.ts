type mapper<I, O> = (input: I) => O;
type updater<T> = (f: ((v: T) => T) | T) => void;
type getterProps<S> =  {[K in keyof S]: S[K]};
export type composeWithState =
  <P extends {}, S extends {}>(initialState: S | mapper<P, S>, options?: any) =>
    React.ComponentType<P> => React.ComponentType<P & getterProps<S> & any>;
