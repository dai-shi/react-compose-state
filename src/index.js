import React from 'react';

const isFunction = fn => (typeof fn === 'function');
const capitalize = str => (str.charAt(0).toUpperCase() + str.slice(1));

export const composeWithState = (initialState, options = {}) => BaseComponent => (
  class extends React.PureComponent {
    constructor(props) {
      super(props);
      if (isFunction(initialState)) {
        this.state = initialState(props);
      } else {
        this.state = initialState;
      }
      const { setters = {} } = options;
      this.state = this.state || {};
      this.stateSetters = {};
      Object.keys(this.state).forEach((key) => {
        const setter = setters[key] || `set${capitalize(key)}`;
        this.stateSetters[setter] = (val) => {
          this.setState({ [key]: val });
        };
      });
    }
    render() {
      return (<BaseComponent {...this.props} {...this.state} {...this.stateSetters} />);
    }
  }
);
