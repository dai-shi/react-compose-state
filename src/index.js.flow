declare type InitialState = Object;

declare type ComposeOptions = {
  setters: Object
};

declare function composeWithState(initialState: InitialState, options?: ComposeOptions): ((props: Object) => any)
