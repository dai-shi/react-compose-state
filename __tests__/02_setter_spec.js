/* eslint-env jest */

import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { composeWithState } from '../src/index';

describe('setter spec', () => {
  it('should update state with default setter', () => {
    const BaseComponent = ({ text, setText }) => (
      <div>
        <h1>{text}</h1>
        <button onClick={() => setText('replaced')}>button</button>
      </div>
    );
    BaseComponent.propTypes = {
      text: PropTypes.string.isRequired,
      setText: PropTypes.func.isRequired,
    };
    const initialState = { text: 'hello' };
    const ComposedComponent = composeWithState(initialState)(BaseComponent);

    const wrapper = mount(<ComposedComponent />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('button').simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should update state with custom setter', () => {
    const BaseComponent = ({ text, updateText }) => (
      <div>
        <h1>{text}</h1>
        <button onClick={() => updateText('replaced')}>button</button>
      </div>
    );
    BaseComponent.propTypes = {
      text: PropTypes.string.isRequired,
      updateText: PropTypes.func.isRequired,
    };
    const initialState = { text: 'hello' };
    const options = { setters: { text: 'updateText' } };
    const ComposedComponent = composeWithState(initialState, options)(BaseComponent);

    const wrapper = mount(<ComposedComponent />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.find('button').simulate('click');
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
