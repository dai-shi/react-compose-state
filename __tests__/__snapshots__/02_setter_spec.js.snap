// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setter spec should update state with custom setter 1`] = `
<_class>
  <BaseComponent
    text="hello"
    updateText={[Function]}
  >
    <div>
      <h1>
        hello
      </h1>
      <button
        onClick={[Function]}
      >
        button
      </button>
    </div>
  </BaseComponent>
</_class>
`;

exports[`setter spec should update state with custom setter 2`] = `
<_class>
  <BaseComponent
    text="replaced"
    updateText={[Function]}
  >
    <div>
      <h1>
        replaced
      </h1>
      <button
        onClick={[Function]}
      >
        button
      </button>
    </div>
  </BaseComponent>
</_class>
`;

exports[`setter spec should update state with default setter 1`] = `
<_class>
  <BaseComponent
    setText={[Function]}
    text="hello"
  >
    <div>
      <h1>
        hello
      </h1>
      <button
        onClick={[Function]}
      >
        button
      </button>
    </div>
  </BaseComponent>
</_class>
`;

exports[`setter spec should update state with default setter 2`] = `
<_class>
  <BaseComponent
    setText={[Function]}
    text="replaced"
  >
    <div>
      <h1>
        replaced
      </h1>
      <button
        onClick={[Function]}
      >
        button
      </button>
    </div>
  </BaseComponent>
</_class>
`;
